cpp/src/hash/hashing.cu

```diff
@@ -100,61 +100,73 @@ hash_partition_table(table_view const& input,
     // Compute which partition each row belongs to using bitwise partitioner
     auto partitioner = bitwise_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
-
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   } else {
     // Compute which partition each row belongs to using modulo partitioner
     auto partitioner = modulo_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   }
 
-  // Scatter input rows to output partitions given the partition map
-  auto const partition_map = column_view(data_type(INT32), num_rows,
-    row_partition_numbers.data().get());
-  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
+  // Initialize gather maps and offsets to sequence
+  auto gather_map = rmm::device_vector<size_type>(num_rows);
+  auto d_offsets = rmm::device_vector<size_type>(num_rows);
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    gather_map.begin(), gather_map.end());
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    d_offsets.begin(), d_offsets.end());
 
-  // Pad with empty tables if we have less than num_partitions
-  if (output.size() < static_cast<size_t>(num_partitions)) {
-    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
-      [&input]() { return experimental::empty_like(input); });
-  }
+  // Sort sequence using partition map as key to generate gather maps
+  thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), gather_map.begin());
 
-  return output;
+  // Reduce unique partitions to extract gather map offsets from sequence
+  auto end = thrust::unique_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), d_offsets.begin());
+
+  auto output = experimental::detail::gather(input, gather_map.begin(), gather_map.end(),
+    false, false, false, mr, stream);
+
+  // Copy offsets from device to vector, padded to num_partitions
+  std::vector<size_type> offsets(num_partitions, input.num_rows());
+  thrust::copy(d_offsets.begin(), end.second, offsets.begin());
+
+  return std::make_pair(std::move(output), std::move(offsets));
```

Comment:
I wasn't sure about this. Is make_pair smart enough to not invoke copies here? I'll have to do some investigation

Reply:
100% guaranteed that a copy is not made.

99.9% guaranteed that the compiler will just use return-value optimization and just use the returned object directly within the function. (100% guaranteed in C++17 conformant compilers where RVO is mandatory). 

In the 0.01% chance the compiler _doesn't_ do RVO, then it will use the move constructor (if available).

By using `std::move` here, you _force_ the compiler to use the move ctor, which is technically more expensive than RVO.

See
https://www.fluentcpp.com/2016/11/28/return-value-optimizations/

https://en.cppreference.com/w/cpp/language/copy_elision

Reply:
Hm, I actually get a compiler error if I try to do this. It seems that `unique_ptr` must be moved into `make_pair`?

Reply:
After digging into it some more, it appears that if you don't move into make_pair, it just does copy construction (which is why it fails for the unique_ptr). Regardless of which of the above you do, it _does_ appear that the pair object itself is eligible for RVO.

Reply:
Yeah, I was totally wrong. I neglected the fact that `std::pair` is an object that first needs to be constructed before it can be returned.